packages/pkl.lua/lua.pkl

```diff
@@ -0,0 +1,292 @@
+  before: String?
+
+  /// The value to render.
+  value: Any
+
+  /// The text to output after rendering [value].
+  after: String?
+}
+
+/// Renders values as Lua.
+class Renderer extends ValueRenderer {
+  /// The characters to use for indenting output. Defaults to two spaces.
+  indent: String = "  "
+
+  /// Whether to skip rendering properties whose value is [null].
+  ///
+  /// Note that due to language limitations, any entries in a [Dynamic] will be treated as
+  /// properties.
+  omitNullProperties: Boolean = false
+
+  /// The number of elements in a [Mapping] or [Listing] to render them multiline.
+  ///
+  /// The default value of `2` means a collection with a single element will be rendered in the
+  /// inline style, although any nested collections may be rendered in the multiline style.
+  ///
+  /// Note that [Map] and [List] are always rendered in the inline style.
+  multilineThreshold: Int = 2
+
+  /// Value converters to apply before values are rendered.
+  ///
+  /// For more information see [ValueRenderer.converters]. Note that due to language limitations,
+  /// any entries in a [Dynamic] will have converters applied as though the entry was a property.
+  converters: Mapping<(Class|String), (unknown) -> Any>
+
+  extension = "lua"
+
+  function renderValue(value: Any): String =
+    let (path = List("^"))
+    render(convert(value, path), path, 0)
+
+  function renderDocument(value: Any): String =
+    let (path = List("^"))
+    let (value = convert(value, path))
+    if (value is RenderDirective) "\(value.text)\n"
+    else if (value is LuaRenderDirective) "\(value.before ?? "")\(render(value.value, path, 0))\(value.after ?? "")\n"
+    else if (value is Dynamic|Typed)
+      value.toMap().fold("", (acc, k, v) ->
+        let (path = path.add(Prop(k)))
+        let (v = convert(v, path))
+        if (omitNullProperties && v == null && k is String) acc else
+        acc + "\(renderKey(k, "_ENV")) = \(render(v, path, 0))\n")
+      + (if (value is Dynamic) value.toList() else List()).foldIndexed("", (idx, acc, v) ->
+        let (path = path.add(splatKey))
+        let (v = convert(v, path))
+        // remember, Lua indexes are 1-based
+        acc + "_ENV[\(idx+1)] = \(render(v, path, 0))\n")
+    else if (value is Mapping|Map)
+      value.fold("", (acc, k, v) ->
+        let (path = path.add(Key(k)))
+        let (v = convert(v, path))
+        acc + "\(renderKey(k, "_ENV")) = \(render(v, path, 0))\n")
+    else throw("The top-level value of a Lua document must have type `Typed`, `Dynamic`, `Mapping`, or `Map`, but got type `\(value.getClass())`")
+
+  // ## Misc helpers ##
+
+  // Returns a [Prop] for string keys, otherwise [splatKey].
+  local function Prop(k: Any): Prop|Key =
+    if (k is String) new Prop { name = k } else splatKey
+
+  // Returns a [Key] for string keys, otherwise [splatKey]
+  local function Key(k: Any): Key =
+    if (k is String) new Key { key = k } else splatKey
```

Comment:
[nit] place these functions above their class definitions.

E.g.

```
// Returns a [Prop] for string keys, otherwise [splatKey].
local const function Prop(k: Any): Prop|Key =
  if (k is String) new Prop { name = k } else splatKey

local class Prop {
  name: String
}
```